autoconnect/autoconnect-ws/src/handler.rs

```diff
@@ -0,0 +1,291 @@
+use std::sync::Arc;
+
+use actix_ws::{CloseReason, Message};
+use futures::{channel::mpsc, stream::StreamExt};
+use tokio::{
+    select,
+    time::{interval, timeout},
+};
+
+use autoconnect_common::protocol::{ServerMessage, ServerNotification};
+use autoconnect_settings::AppState;
+use autoconnect_ws_sm::{UnidentifiedClient, WebPushClient};
+
+use crate::{
+    error::WSError,
+    session::{Session, SessionImpl},
+};
+
+/// WebPush WebSocket handler Task
+pub async fn webpush_ws(
+    session: actix_ws::Session,
+    msg_stream: actix_ws::MessageStream,
+    app_state: Arc<AppState>,
+    ua: String,
+) {
+    let client = UnidentifiedClient::new(ua, app_state);
+    let mut session = SessionImpl::new(session.clone());
+    let close_reason = _webpush_ws(client, &mut session, msg_stream)
+        .await
+        .unwrap_or_else(|e| {
+            Some(CloseReason {
+                code: e.close_code(),
+                // TODO: clip this String to =~ 100, WS control frames (such as
+                // close) are limited to 125 bytes
+                description: Some(e.to_string()),
+            })
+        });
+    trace!("webpush_ws: close_reason: {:#?}", close_reason);
+    let _ = session.close(close_reason).await;
+}
+
+/// WebPush WebSocket handler
+///
+/// Transistions the client between the `UnidentifiedClient` (waiting for a
+/// Hello) state and the identified full `WebPushClient`.
+///
+/// Manages:
+///
+///- I/O to and from the client: incoming `ClientMessage`s from the WebSocket
+/// connection and `ServerNotification`s from the `ClientRegistry` (triggered
+/// by `autoendpoint`), and in turn outgoing `ServerMessage`s written to the
+/// WebSocket connection in response to those events.
+/// - the lifecycle/cleanup of the Client
+async fn _webpush_ws(
```

Comment:
Yeah, let's rename this or `webpush_ws` (maybe call that `webpush_init` or something?)

Reply:
This top level webpush_ws as is isn't very testable anyway so I just moved the spawn into it and renamed it accordingly.